### PR TITLE
planner: fix statement-optimize not work in `TryFastPlan`

### LINKED ISSUE #18042: statement-level optimize hint will be ignored when `tryFastPlan` works
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
Found in https://github.com/pingcap/tidb/issues/17990.
Code link: https://github.com/pingcap/tidb/blob/master/planner/optimize.go#L43-L56

### 2. What did you expect to see? (Required)
The statement-level optimize hint can also work when tryFastPlan works

### 4. Affected version (Required)

v4.0.0

### 5. Root Cause Analysis

As the `Minimal reproduce step` shows
## SIG slack channel

 [#sig-planner](https://slack.tidb.io/invite?team=tidb-community&channel=sig-planner&ref=high-performance)

## Score

- 300

## Mentor

* @wshwsh12

### PR DESCRIPTION
**Pull request by Patrick0308**
### What problem does this PR solve?

Issue Number: close #18042 

Problem Summary:

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- planner: fix statement-optimize not work in `TryFastPlan`

### CODE REVIEW

**Review by wshwsh12 — COMMENTED** [review]
Please add some test cases. Rest LGTM.

**Comment by XuHuaiyu** [comment]
@Patrick0308 Friendly ping, it'll be nice if you can address the comment

**Comment by Patrick0308** [comment]
@XuHuaiyu Sorry for that it is not finished yet. I will fix this problem right away

**Comment by Patrick0308** [comment]
@wshwsh12 PTAL

**Review by wshwsh12 — COMMENTED** [review]
LGTM

File: planner/core/point_get_plan_test.go [hunk 1]
```diff
@@ -600,3 +600,14 @@ func (s *testPointGetSuite) TestCBOShouldNotUsePointGet(c *C) {
...
 	}
 }
+
+func (s *testPointGetSuite) TestIssue18042(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table t")
+	tk.MustExec("create table t(a int, b int, c int, primary key(a));")
+	tk.MustExec("insert into t values (1, 1, 1), (2, 2, 2), (3, 3, 3), (4, 4, 4)")
+	tk.MustExec("SELECT /*+ MAX_EXECUTION_TIME(100)*/ * FROM t where a = 1;")
```
Comment (zz-jason) [inline planner/core/point_get_plan_test.go:610]: please add more test for other hints, like `use_index()`

**Review by zz-jason — CHANGES_REQUESTED** [review]

File: planner/core/point_get_plan_test.go [hunk 2]
```diff
@@ -600,3 +600,14 @@ func (s *testPointGetSuite) TestCBOShouldNotUsePointGet(c *C) {
 		res.Check(testkit.Rows(output[i].Res...))
 	}
 }
+
+func (s *testPointGetSuite) TestIssue18042(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table t")
```
Comment (qw4990) [inline planner/core/point_get_plan_test.go:607]: Please use `drop table if exists t` which is more robust.

Comment (qw4990) [reply planner/core/point_get_plan_test.go]: addressed.

File: planner/core/point_get_plan_test.go (hunk 1 above)
Comment (qw4990) [reply planner/core/point_get_plan_test.go]: Index hints like `use_index` seem trivial to fast plans like point-get, while runtime hints like exec-time or mem-quota are more important.
I added mem-quota into this test, PTAL.

Comment (rebelice) [reply planner/core/point_get_plan_test.go]: Maybe we need to determine the behavior of `use_index()` with `point get`. For example, when we specify an index that cannot use `point get`, what will the optimizer do?

**Review by winoros — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

File: planner/core/point_get_plan_test.go (hunk 1 above)
Comment (qw4990) [reply planner/core/point_get_plan_test.go]: Yes~ `use_index()` can be applied to `point get`, but I think it's trivial.
For example, for the query `select a from t where a=1`, using `index(a)` or `index(a, b)` as its access path in a `point get` has the same efficiency.
And `point get` doesn't support `use_index()` now, so we have to introduce more lines of code into `point get` if we want to support and test it.
On second thought, I think it's better to doesn't change it and keep the status quo.

**Review by zz-jason — APPROVED** [review]
LGTM

**Pull request merged**